File: src/main/java/io/github/petrknap/persistence/zoneddatetime/UtcWithTimezone.java
package io.github.petrknap.persistence.zoneddatetime;

import jakarta.persistence.Column;
import jakarta.persistence.Embeddable;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZonedDateTime;

/**
 * Stores zoned date-time as `utc` date-time with `timezone` identifier
 */
@Embeddable
public final class UtcWithTimezone extends Utc<UtcWithTimezone>
{
    @Column(length = 64, nullable = true)
    private @Nullable String timezone;

    public UtcWithTimezone(@NotNull ZonedDateTime zonedDateTime)
    {
        super(zonedDateTime);
        timezone = zonedDateTime.getZone().getId();
    }

    private UtcWithTimezone()
    {
        super();
    }

    public static @Nullable UtcWithTimezone fromFormattedValues(
            @Nullable CharSequence utcDateTime,
            @NotNull String dateTimeFormat,
            @Nullable String timezone
    ) {
        return fromValues(
                utcDateTime != null ? DateTimeUtils.parseAsLocalDateTime(utcDateTime, dateTimeFormat) : null,
                timezone != null ? ZoneId.of(timezone) : null
        );
    }

    public static @Nullable UtcWithTimezone fromValues(
            @Nullable LocalDateTime utcDateTime,
            @Nullable ZoneId timezone
    ) {
        ZonedDateTime zonedDateTime = ZonedDateTimePersistence.computeZonedDateTime(
                utcDateTime,
                timezone
        );

        return zonedDateTime != null ? new UtcWithTimezone(zonedDateTime) : null;
    }

    public @NotNull ZoneId getTimezone()
    {
        return ZoneId.of(getTimezone(true));
    }

    public @NotNull String getTimezone(boolean formatted)
    {
        if (!formatted) {
            throw new IllegalArgumentException("Argument formatted must be true");
        }
        if (timezone == null) {
            thisInstanceShouldBeNull();
        }
        return timezone;
    }

    public @NotNull ZonedDateTime toZonedDateTime()
    {
        return ZonedDateTimePersistence.computeZonedDateTime(
                getUtcDateTime(),
                getTimezone()
        );
    }
}

File: src/main/php/UtcWithTimezone.php
<?php

declare(strict_types=1);

namespace PetrKnap\ZonedDateTimePersistence;

use DateTimeImmutable;
use DateTimeInterface;
use DateTimeZone;
use Doctrine\ORM\Mapping as ORM;

/**
 * Stores zoned date-time as `utc` date-time with `timezone` identifier
 */
#[ORM\Embeddable]
final class UtcWithTimezone extends Utc
{
    #[ORM\Column(length: 64, nullable: true)]
    protected string|null $timezone;

    public function __construct(DateTimeInterface $zonedDateTime)
    {
        parent::__construct($zonedDateTime);
        $this->timezone = $zonedDateTime->getTimezone()->getName();
    }

    public static function fromFormattedValues(
        string|null $utcDateTime,
        string $dateTimeFormat,
        string|null $timezone,
    ): UtcWithTimezone|null {
        return self::fromValues(
            $utcDateTime !== null ? DateTimeUtils::parseAsLocalDateTime($utcDateTime, $dateTimeFormat) : null,
            $timezone !== null ? new DateTimeZone($timezone) : null,
        );
    }

    public static function fromValues(
        DateTimeInterface|null $utcDateTime,
        DateTimeZone|null $timezone,
    ): UtcWithTimezone|null {
        $zonedDateTime = ZonedDateTimePersistence::computeZonedDateTime(
            $utcDateTime,
            timezone: $timezone,
        );

        return $zonedDateTime != null ? new UtcWithTimezone($zonedDateTime) : null;
    }

    /**
     * @return ($formatted is false ? DateTimeZone : string)
     */
    public function getTimezone(bool $formatted = false): DateTimeZone|string
    {
        if ($this->timezone === null) {
            $this->thisInstanceShouldBeNull();
        }
        return $formatted ? $this->timezone : new DateTimeZone($this->timezone);
    }

    public function toZonedDateTime(): DateTimeImmutable
    {
        return ZonedDateTimePersistence::computeZonedDateTime(
            $this->getUtcDateTime(),
            timezone: $this->getTimezone(),
        );
    }
}
